content/developer/reference/javascript/patching_code.rst

```diff
@@ -0,0 +1,194 @@
+  cannot call ``this._super`` after an ``await``, because it may or may not be
+  the function that you expect.  The correct way to do that is to keep a reference
+  to the initial ``_super`` method:
+
+  .. code-block:: javascript
+
+    patch(object, "async _super patch", {
+      async myAsyncFn() {
+        const _super = this._super;
+        await Promise.resolve();
+        await _super(...arguments);
+        // await this._super(...arguments); // this._super is undefined.
+      },
+    });
+
+
+Getters and setters are supported too:
+
+.. code-block:: javascript
+
+    patch(object, "getter/setter patch", {
+      get number() {
+        return this._super() / 2;
+      },
+      set number(value) {
+        this._super(value * 2);
+      },
+    });
+
+Patching a javascript class
+===========================
+
+The ``patch`` function is designed to work with anything: object or ES6 class.
+
+However, since javascript classes work with the prototypal inheritance, when
+one wishes to patch a standard method from a class, then we actually need to patch
+the `prototype`:
+
+.. code-block:: javascript
+
+  class MyClass {
+    static myStaticFn() {...}
+    myPrototypeFn() {...}
+  }
+
+  // this will patch static properties!!!
+  patch(MyClass, "static patch", {
+    myStaticFn() {...},
+  });
+
+  // this is probably the usual case: patching a class method
+  patch(MyClass.prototype, "prototype patch", {
+    myPrototypeFn() {...},
+  });
+
+
+Also, Javascript handle the constructor in a special native way which makes it
+impossible to be patched. The only workaround is to call a method in the original
+constructor and patch that method instead:
+
+.. code-block:: javascript
+
+  class MyClass {
+    constructor() {
+      this.setup();
+    }
+    setup() {
+      this.number = 1;
+    }
+  }
+
+  patch(MyClass.prototype, "constructor", {
+    setup() {
+      this._super(...arguments);
+      this.doubleNumber = this.number * 2;
+    },
+  });
+
+.. warning::
+
+    It is impossible to patch directly the `constructor` of a class!
+
+Patching a component
+====================
+
+Components are defined by javascript classes, so all the information above still
+holds.  For these reasons, Owl components should use the `setup` method, so they
+can easily be patched as well (see the section on :ref:`best practices<javascript/owl_best_practices>`.
+
+.. code-block:: javascript
+
+  patch(MyComponent.prototype, "my patch", {
+    setup() {
+      useMyHook();
+    },
+  });
+
+Removing a patch
+================
+
+The `patch` function has a counterpart, `unpatch`, also located in `@web/core/utils/patch`.
+
+.. js:function:: unpatch(obj, patchName)
+
+    :param Object obj: object that should be unpatched
+    :param string patchName: string describing the patch that should be removed
+
+    Removes an existing patch from an object `obj`. This is mostly useful for
+    testing, when there may be various active patches that interfere with tests.
```

Comment:
not sure about this, we don't really remove patches in tests, because we can't re-apply them later. However, unpatch is very useful in tests, because we often patch something at the beginning of a test, and unpatch it at the end